core/services/log/registrations.go

```diff
@@ -0,0 +1,283 @@
+package log
+
+import (
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+)
+
+// How it works in general:
+//
+// 1. Adding and removing listeners updates the highestNumConfirmations - a number tracking what's the current highest NumConfirmations globally
+//
+// 2. All received logs are kept in an array and deleted ONLY after they are outside the confirmation range for all subscribers
+// (when given log height is lower than (latest height - highestNumConfirmations) ) -> see: pool.go
+//
+// 3. The logs are attempted to be sent after every new head arrival:
+// 		Each stored log is then checked against every matched listener and is sent unless:
+//    A) is too young for that listener
+//    B) the corresponding block height is known to be already processed for that listener
+// In the normal case, each log will be only processed once, and then its corresponding head will be remembered, so it's not double-sent
+//
+// After processing the whole batch of logs considered for sending, the per-listener metadata is updated in applyListenerInfoUpdates.
+// If a re-org happens, the stored lowestAllowedBlockNumber (per-listener) is re-set,
+// so the logs from that chain are then considered unprocessed, and will be sent again.
+//
+type (
+	registrations struct {
+		registrations map[common.Address]map[common.Hash]map[Listener]*listenerMetadata // contractAddress => logTopic => Listener
+		decoders      map[common.Address]AbigenContract
+
+		// highest 'num_confirmations' per all listeners, used to know when to delete old logs
```

Comment:
This is basically `max(listeners.map(l => l.num_confirmations))` right?

Reply:
yes, will clarify